saml/src/main/java/com/linecorp/armeria/saml/SamlDecorator.java

```diff
@@ -0,0 +1,155 @@
+    private final Credential signingCredential;
+    private final CompositeAuthorizer<HttpRequest> authorizers;
+    private final SamlAuthManager authManager;
+    private final SamlRequestIdManager requestIdManager;
+
+    SamlDecorator(SamlServiceProviderConfigurator configurator,
+                  Service<HttpRequest, HttpResponse> delegate) {
+        super(delegate);
+        this.configurator = configurator;
+
+        myEntityId = configurator.myEntityId();
+        signingCredential = configurator.signingCredential();
+        authorizers = configurator.authorizers();
+        authManager = configurator.authManager();
+        requestIdManager = configurator.requestIdManager();
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        return HttpResponse.from(authorizers.authorize(ctx, req).handleAsync((result, cause) -> {
+            if (cause == null && result) {
+                // Already authenticated.
+                try {
+                    return delegate().serve(ctx, req);
+                } catch (Exception e) {
+                    return Exceptions.throwUnsafely(e);
+                }
+            }
+            // Find an identity provider first where the request is to be sent to.
+            return HttpResponse.from(configurator.idpConfigSelector()
+                                                 .select(configurator, ctx, req).thenApply(idp -> {
+                        if (idp == null) {
+                            return HttpResponse.of(HttpStatus.BAD_REQUEST);
+                        }
+
+                        final AuthnRequest request = createAuthRequest(idp);
+                        final String relayState = authManager.getRelayState(ctx, req);
+                        // Support HTTP-Redirect and HTTP-POST binding protocols when sending
+                        // an authentication request.
+                        final SamlEndpoint endpoint = idp.ssoEndpoint();
+                        if (endpoint.bindingProtocol().isHttpRedirect()) {
+                            return responseWithLocation(toRedirectionUrl(
+                                    request, endpoint.urlAsString(), PARAM_SAML_REQUEST,
+                                    signingCredential, configurator.signatureAlgorithm(), relayState));
+                        } else {
+                            final String value = signAndBase64Encode(request, signingCredential,
+                                                                     configurator.signatureAlgorithm());
+                            final HttpData body = getHtmlForm(endpoint.urlAsString(),
+                                                              PARAM_SAML_REQUEST, value, relayState);
+                            return HttpResponse.of(HttpStatus.OK, MediaType.HTML_UTF_8, body);
+                        }
+                    }));
+        }, ctx.blockingTaskExecutor()));
```

Comment:
I don't think we're supposed to force `delegate().serve` onto the blocking executor.

Reply:
Good point. Will not use the blocking executor here. Thanks!